File: leetsync/2665-minimum-time-to-repair-cars/minimum-time-to-repair-cars.go
func repairCars(ranks []int, cars int) int64 {
	maxRank := math.MinInt32
	for _, rank := range ranks {
		maxRank = max(maxRank, rank)
	}
	var left int64 = 1
	var right int64 = int64(maxRank * cars * cars)
	var result int64 = 0
	for left <= right {
		mid := int64(left + (right-left)/2)
		if isPossibleToRepair(ranks, cars, mid) {
			result = mid
			right = mid - 1
		} else {
			left = mid + 1
		}
	}
	return result
}
func isPossibleToRepair(ranks []int, cars int, mid int64) bool {
	var repairdCars int64 = 0
	for _, rank := range ranks {
		repairdCars += int64(math.Sqrt(float64(mid / int64(rank))))
	}
	return repairdCars >= int64(cars)
}

File: leetsync/2665-minimum-time-to-repair-cars/minimum-time-to-repair-cars.java
class Solution {
    public boolean isPossibleToRepair(int[] ranks, int cars, long mid) {
        long repairdCars = 0L;
        int n = ranks.length;
        for (int rank : ranks) {
            repairdCars += Math.sqrt(mid / rank);
        }
        return repairdCars >= cars;
    }

    public long repairCars(int[] ranks, int cars) {
        int n = ranks.length;
        int maxRank = Integer.MIN_VALUE;
        for (int rank : ranks) {
            maxRank = Math.max(rank, maxRank);
        }
        long start = 1, end = 1L * maxRank * cars * cars, result = 0;
        while (start <= end) {
            long mid = start + (end - start) / 2;
            if (isPossibleToRepair(ranks, cars, mid)) {
                result = mid;
                end = mid - 1;
            } else {
                start = mid + 1;
            }
        }
        return result;
    }
}